tpu-client-next/src/connection_worker.rs

```diff
@@ -0,0 +1,220 @@
+    connection: ConnectionState,
+    send_txs_stats: SendTransactionStats,
+}
+
+impl ConnectionWorker {
+    pub fn new(
+        endpoint: Endpoint,
+        peer: SocketAddr,
+        command_receiver: mpsc::Receiver<TransactionBatch>,
+    ) -> Self {
+        Self {
+            endpoint,
+            peer,
+            transactions_receiver: command_receiver,
+            connection: ConnectionState::NotSetup,
+            send_txs_stats: SendTransactionStats::default(),
+        }
+    }
+
+    /// Starts the main loop of the [`ConnectionWorker`]. This method manages the
+    /// connection to the peer and handles state transitions. It runs
+    /// indefinitely until the connection is closed or an unrecoverable error
+    /// occurs.
+    pub async fn run(&mut self) {
+        loop {
+            match &self.connection {
+                ConnectionState::Closing => {
+                    break;
+                }
+                ConnectionState::NotSetup => {
+                    self.create_connection(0).await;
+                }
+                ConnectionState::Active(connection) => {
+                    let Some(transactions) = self.transactions_receiver.recv().await else {
+                        info!("Transactions sender has been dropped.");
+                        self.connection = ConnectionState::Closing;
+                        continue;
+                    };
+                    self.send_transactions(connection.clone(), transactions)
+                        .await;
+                }
+                ConnectionState::Retry(num_reconnects) => {
+                    if *num_reconnects > MAX_RECONNECT_ATTEMPTS {
+                        error!(
+                            "Didn't manage to establish connection: reach max reconnect attempts."
+                        );
+                        self.connection = ConnectionState::Closing;
+                        continue;
+                    }
+                    sleep(RETRY_SLEEP_INTERVAL).await;
+                    self.reconnect(*num_reconnects).await;
+                }
+            }
+        }
+    }
+
+    /// Retrieves the statistics for transactions sent by this worker.
+    pub fn get_transaction_stats(&self) -> &SendTransactionStats {
+        &self.send_txs_stats
+    }
+
+    /// Sends a batch of transactions using the provided `connection`. Each
+    /// transaction in the batch is sent over the QUIC streams one at the time,
+    /// which prevents traffic fragmentation and shows better TPS in comparison
+    /// with multistream send. If the batch is determined to be outdated, it
+    /// will be dropped without being sent. In case of error, it doesn't retry
+    /// to send the same transactions again.
+    async fn send_transactions(&mut self, connection: Connection, transactions: TransactionBatch) {
+        let now = timestamp();
+        if now.saturating_sub(transactions.get_timestamp()) > MAX_PROCESSING_AGE_MS {
```

Comment:
This logic could be done over block height, usually transaction is valid for 300 block, during epoch boundary first slot may take few seconds to come, in this case the transactions will be dropped.

Reply:
That would definitely be more precise, but I think there are 2 challenges:

1. The transactions are already serialized into bytes at this point, so extracting blockhash/slot is tricky.
2. Even if we're within the 150 block limit, by the time it gets sent/received/included in a block, we might be outside the window, so we would likely need to introduce some fudge factor anyways.

Reply:
@KirillLykov Do you know if we're hitting this and dumping tx batches during normal operation? I'm assuming it depends on where the bottleneck is between tx generation/scheduling/sending and the backpressure mechanism

Reply:
Discussed these concerns with @godmodegalactus and Groovie|mango and I'm quite convinced that this thing must be implemented if we want to use this client not only for benchmarking cluster.

Reply:
I don't see a simple way to move this check to the caller because it is a channel where caller puts his transactions and he cannot go over it and clean them up. But I think a decent solution is to do some age check optionally (controlled by boolean flag). Because actually there is only one correct way to check the age of tx and because there are only two possible options -- we want this check or not.